pyproject.toml

```diff
@@ -1,6 +1,61 @@
 [build-system]
-requires = ["setuptools", "wheel"]
-build-backend = "setuptools.build_meta"
+requires = ["hatchling>=0.14.0"]
+build-backend = "hatchling.build"
+
+[project]
+name = "pipx"
+description = "Install and Run Python Applications in Isolated Environments"
+readme = "README.md"
+license = "MIT"
+requires-python = ">=3.7"
+keywords = [
+    "pip",
+    "install",
+    "cli",
+    "workflow",
+    "Virtual Environment",
+]
+authors = [
+    { name = "Chad Smith", email = "chadsmith.software@gmail.com" },
+]
+classifiers = [
+    "License :: OSI Approved :: MIT License",
+    "Operating System :: OS Independent",
+    "Programming Language :: Python",
+    "Programming Language :: Python :: 3 :: Only",
```

Comment:
One feature gap compared to setuptools and `setup.cfg` I see here is that this file is now manually maintained and formatted while using `setuptools` one can use the excellent `setup-cfg-fmt` project for it. Does hatch have plans to add formatting support? Without it I'm tempted to keep using setuptools :thinking: Naturally if other maintainers even with that feature gap like setuptools I'm not going to oppose :+1:

Reply:
I _suppose_ I could add a command for that but:

1. Something like that is indeed better suited as a small, separate project that can be used by all.
2. I would personally never use it because I would like to retain comments, and `tomlkit` is extremely buggy.

Are you saying that until there is a `pyproject.toml` linter you're not keen on using PEP 621?

Reply:
Indeed :blush:

Reply:
Setup-cfg-fmt does a lot more than just formatting, it also keeps up the classifiers for new Python versions and cleans them up, sorts dependencies, etc. It also runs in pre-commit, can taplo?

It also strips comments, valid empty (sub) sections, and can’t be disabled for part of a file AFAIK…

Reply:
> tomlkit is extremely buggy.
👍

An eventual tool should not be tied to a build backend, but be a generic tool; probably a pair of tools like taplo & (imaginary atm) pep621-fmt or something. Happy to have it live alongside hatch though. ;)

Reply:
> 2\. I would like to retain comments

I'm personally not a fan of comments in configuration files like these, so for me less of an issue. 😆

Reply:
There are some very good reasons to put comments in sometimes, like https://github.com/scikit-hep/awkward-1.0/blob/6d3c1d23d7ab492d0178b1798e179088a066e00a/pyproject.toml#L90-L157 - probably rare for the project metadata itself, but the other tool fields might need comments (flake8/pylint config being big ones for needed comments). Comments can be overdone, sure, but sometimes they are needed. setup-cfg-fmt doesn't support them for technical reasons, not primarily because they are a bad idea. I usually have managed without comments in setup.cfg, though, sometimes painfully with flake8, but really hope not to leak that issue over into pyproject.toml. :)

Reply:
> I would personally never use it because I would like to retain comments, and tomlkit is extremely buggy.

`tomlkit` has received a lot of improvements over the last 6 months. I personally have contributed several patches and @frostming has become a maintainer and has been doing an amazing job!

As an example, I can safely use it for [ini2toml](https://ini2toml.readthedocs.io/en/latest/) for translating the "AST" of `.ini` files to `.toml`.

So if its being a while that you last checked `tomlkit`, maybe it is a good time to have a look if it improved for your use cases?

Reply:
For what's worth I've been using `tomlkit` for https://github.com/gaborbernat/pyproject-fmt and works well for me 👍

Reply:
Run into this due to the ping, and I found @gaborbernat already made a [pyproject.toml formatter ](https://github.com/gaborbernat/pyproject-fmt), powered by `tomlkit`

Reply:
I tried to use toolkit in cibuildwheel (at least 6 months ago, probably more), and found it too buggy to use; it was a rather weird format, but it was important to keep it, which it failed to do. I ended up just writing a manual writer, since it was a very well structured limited file. Great to hear it is seeing improvements! I know it was not 1.0.0 compliant either last I checked, but is now. :)

And @gaborbernat, @abravalheri, and @ofek have all made TOML formatter/validators. :) `pypyroject-fmt`, `pyproject-validate`, and `validate-pyproject`.